core/src/server/rpc_module.rs

```diff
@@ -923,91 +1051,86 @@ impl SubscriptionSink {
 	/// use anyhow::anyhow;
 	///
 	/// let mut m = RpcModule::new(());
-	/// m.register_subscription("sub", "_", "unsub", |params, mut sink, _| {
+	/// m.register_subscription("sub", "_", "unsub", |params, pending, _| async move {
+	///     let mut sink = pending.accept().await?;
+	///
 	///     let stream = futures_util::stream::iter(vec![Ok(1_u32), Ok(2), Err("error on the stream")]);
 	///     // This will return send `[Ok(1_u32), Ok(2_u32), Err(Error::SubscriptionClosed))]` to the subscriber
-	///     // because after the `Err(_)` the stream is terminated.
+	///     // because after the `Err(_)` then the stream is terminated.
 	///     let stream = futures_util::stream::iter(vec![Ok(1_u32), Ok(2), Err("error on the stream")]);
 	///
-	///     tokio::spawn(async move {
-	///
-	///         // jsonrpsee doesn't send an error notification unless `close` is explicitly called.
-	///         // If we pipe messages to the sink, we can inspect why it ended:
-	///         match sink.pipe_from_try_stream(stream).await {
-	///            SubscriptionClosed::Success => {
+	///     // jsonrpsee doesn't send an error notification unless `close` is explicitly called.
+	///     // If we pipe messages to the sink, we can inspect why it ended:
+	///     match sink.pipe_from_try_stream(|last, next| last.unwrap_or(next), stream).await {
+	///         SubscriptionClosed::Success => {
 	///                let err_obj: ErrorObjectOwned = SubscriptionClosed::Success.into();
-	///                sink.close(err_obj);
-	///            }
-	///            // we don't want to send close reason when the client is unsubscribed or disconnected.
-	///            SubscriptionClosed::RemotePeerAborted => (),
-	///            SubscriptionClosed::Failed(e) => {
-	///                sink.close(e);
-	///            }
+	///                sink.close(err_obj).await;
+	///         }
+	///         // we don't want to send close reason when the client is unsubscribed or disconnected.
+	///         SubscriptionClosed::RemotePeerAborted => (),
+	///         SubscriptionClosed::Failed(e) => {
+	///           sink.close(e).await;
 	///         }
-	///     });
+	///     }
 	///     Ok(())
 	/// });
 	/// ```
-	pub async fn pipe_from_try_stream<S, T, E>(&mut self, mut stream: S) -> SubscriptionClosed
+	pub async fn pipe_from_try_stream<'a, S, T, F, E>(&'a mut self, f: F, stream: S) -> SubscriptionClosed
 	where
+		F: Fn(Option<PendingOrFuture<'a>>, PendingOrFuture<'a>) -> PendingOrFuture<'a>,
```

Comment:
I also guess that this method doesn't need any "special" access to Self; it just uses the public interface to send messages. So I suppose it owuld be a separate thing in some utility module or whatever if we wanted and we don't need to complicate the "core" stuff by having it here if that makes life any easier for us!

It's also the only place `SubscriptionClosed` is returned I think? And maybe it would be nice just to pull that and this out into a separate place :)

Reply:
I would be happy to remove this API completely but `util` is probably a good compromise.

This API is still super hard to understand and use, especially with the `SubscriptionClosed` result whether to manually send a customized close message or not.

Reply:
> The function would return msg; something to try sending on the stream. This might be built be combining the two arguments together (aggregating messages if the first hasn't managed to send yet) or it might just always ignore the first message or whatever.

> Now that these PendingOrFuture args are here, there's no way to aggregate messages; only pick between ignoring one or the other

I'm not sure we should really support aggregating messages, if the subscription is defined with a `Item = usize` when aggregating would imply that subscription notification would like like:

```
# "not aggregated"
{"jsonrpc":"2.0","method":"my_sub","params":{"subscription":1882468350819392,"result":1}}
# aggregated as `Vec`, this will break the client code (potential footgun)
{"jsonrpc":"2.0","method":"my_sub","params":{"subscription":1882468350819392,"result":[1, 2]}}
# aggregated as sum
{"jsonrpc":"2.0","method":"my_sub","params":{"subscription":1882468350819392,"result":3}}
```

We could actually require` |Option<T>, T| -> T` to circumvent that

However, in order to support aggregating we are forced to use `try_send` and perhaps use `the reserve API` if there is any unsent message waiting for a slot. Instead of having a timeout to re-transmit periodically.

Reply:
> We could actually require |Option<T>, T| -> T to circumvent that
> 
> However, in order to support aggregating we are forced to use try_send and perhaps use the reserve API if there is any unsent message waiting for a slot. Instead of having a timeout to re-transmit periodically.

Yeah, `|Option<T>, T| -> T` is what I had in mind. And perhaps there aren't any real use cases for aggregating indeed! I'd be in favoru of just removing this API completely because it's obviously quite complicated, and just adding back what we might find useful as utils!

Reply:
Well, you could keep the original value around until it has sent I suppose? So you serialize, try to send, but keep the non-serialized version in case you need to provide it to the callback :)